pyvisfile/xdmf/__init__.py

```diff
@@ -0,0 +1,1119 @@
+        """
+
+        super().__init__(parent, "Time", {
+            "Value": value,
+            })
+
+# }}}
+
+
+# {{{ domain
+
+class Domain(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(self, *,
+            name: Optional[str] = None,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        """
+
+        super().__init__(parent, "Domain", {
+            "Name": name,
+            })
+
+# }}}
+
+
+# {{{ information
+
+class Information(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(self, *,
+            name: str,
+            value: str,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        """
+
+        super().__init__(parent, "Information", {
+            "Name": name,
+            "Value": value,
+            })
+
+# }}}
+
+
+# {{{ include
+
+class XInclude(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(self, *,
+            href: Optional[str],
+            xpointer: Optional[str] = None,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        :param xpointer: path inside the file represented by *href*.
+        """
+        if xpointer is not None:
+            xpointer = f"xpointer({xpointer})"
+
+        super().__init__(parent, "xi:include", {
+            "href": href,
+            "xpointer": xpointer,
+            })
+
+# }}}
+
+# }}}
+
+
+# {{{ data arrays
+
+def _ndarray_to_string(ary):
+    if not isinstance(ary, np.ndarray):
+        raise TypeError(f"expected an 'ndarray', got '{type(ary).__name__}'")
+
+    ntype = _numpy_to_xdmf_number_type(ary.dtype)
+    if ntype == DataItemNumberType.Int or ntype == DataItemNumberType.UInt:
+        fmt = "%d"
+    elif ntype == DataItemNumberType.Float:
+        fmt = "%.16e" if ary.dtype.itemsize == 8 else "%.8e"
+    else:
+        raise ValueError(f"unsupported dtype: '{ary.dtype}'")
+
+    import io
+    bio = io.BytesIO()
+    np.savetxt(bio, ary, fmt=fmt)
+
+    return "\n" + bio.getvalue().decode()
+
+
+def _geometry_type_from_points(points):
+    dims = points.shape[-1]
+
+    if len(points.components) == 1:
+        if dims == 2:
+            return GeometryType.XY
+        elif dims == 3:
+            return GeometryType.XYZ
+        else:
+            raise ValueError(f"unsupported dimension: '{dims}'")
+    else:
+        if dims == 2:
+            return GeometryType.VXVY
+        elif dims == 3:
+            return GeometryType.VXVYVZ
+        else:
+            raise ValueError(f"unsupported dimension: '{dims}'")
+
+
+class DataArray:
+    r"""An array represented as a list of :class:`DataItem`\ s.
```

Comment:
Kind of confused on the distinction between `DataArray` and `DataItem`. Halp?

Reply:
Well, at least in my head,
* The `DataItem` is just the low-level XDMF/XML tag description.
* The `DataArray` is meant to denote some sort of array data, which can have multiple `DataItems` (e.g. object array).

Reply:
For the object array part, I never managed to get it to work. There's an example in the tests with a `vorticity` variable that's an object array, but Paraview refuses to see it at all.

That's why in inducer/meshmode#98 I just gave up and `np.stack`ed any object arrays before writing them :(